most.1
.\"--------------------------------------------------------------------
.\" UX - UNIX macro, to append \(rg to the first usage
.\"--------------------------------------------------------------------
.de UX
.ie \\n(UX \s-1UNIX\s0\\$1
.el      \{\s-1UNIX\s0\\$1\*(Rg
.          nr UX 1
.\}
..
.\"--------------------------------------------------------------------
.\" Ds - start indented display
.\" De - end indented display
.\"--------------------------------------------------------------------
.de Ds
.nf
'in +0.5i
..
.de De
'in
.fi
..
.TH MOST 1 "May 1999"
.SH NAME
most \- browse or page through a text file
.SH SYNOPSIS
.B most
.RB [ \-1bCcdMstuvwz ]
.RB [ +\fIlineno\fB ]
.RB [ +c ]
.RB [ +d ]
.RB [ +s ]
.RB [ +u ]
.RB [ +/\fIstring\fB ]
.RI [ filename ...]
.SH DESCRIPTION
.I most
is a paging program that displays,
one windowful at a time,
the contents of a file on a terminal.
It pauses after each windowful and prints on the window status line
the screen the file name, current line number,
and the percentage of the file so far displayed.
.PP
Unlike other paging programs,
.I most
is capable of displaying an arbitrary number of windows
as long as each window occupies at least two screen lines.
Each window may contain the same file or a different file.
In addition, each window has its own mode.
For example, one window may display a file with its lines wrapped while
another may be truncating the lines.
Windows may be `locked'
together in the sense that if one of the locked windows scrolls,
all locked windows will scroll.
.I most
is also capable of ignoring lines
that are indented beyond a user specified value.
This is useful when viewing computer programs to pick out
gross features of the code.
See the
.RB ` :o '
command for a description of this feature.
.PP
In addition to displaying ordinary text files,
.I most
can also display binary files as well as files with arbitrary ascii
characters.
When a file is read into a buffer,
.I most
examines the first 32 bytes of
the file to determine if the file is a binary file and then switches
to the appropriate mode.
However, this feature may be disabled with the
.B \-k
option.
See the description of the
.BR \-b ,
.BR \-k ,
.BR \-v ,
and
.B \-t
options for further details.
.PP
Text files may contain combinations of underscore and backspace
characters causing a printer to underline or overstrike.
When
.I most
recognizes this,
it inserts the appropriate escape sequences to
achieve the desired effect.
In addition,
some files cause the printer to overstrike some characters
by embedding carriage return characters in the middle of a line.
When this occurs,
.I most
displays the overstruck character with a bold
attribute.
This feature facilitates the reading of
.UX
man pages or a document produced by
.IR runoff .
In particular,
viewing this document with
.I most
should illustrate this behavior provided that the
underline characters have not been stripped.
This may be turned off with the
.B \-v
option.
.PP
By default, lines with more characters than the terminal width are
not wrapped but are instead truncated.
When truncation occurs, this
is indicated by a `$' in the far right column of the terminal
screen.
The RIGHT and LEFT arrow keys may be used to view lines
which extend past the margins of the screen.
The
.B \-w
option may be used to override this feature.
When a window is wrapped,
the character `\e' will appear at the right edge of the window.
.PP
Commands are listed below.
.SH OPTIONS
.TP
.B \-1
VT100 mode.  This is meaningful only on VMS systems.  This option should be
used if the terminal is strictly a VT100.  This implies that the terminal
does not have the ability to delete and insert multiple lines.  VT102s and
above have this ability.
.TP
.B \-b
Binary mode.
Use this switch when you want to view files
containing 8 bit characters.
.I most
will display the file 16 bytes per line in hexadecimal notation.
A typical line looks like:
.IP
.Ds
01000000 40001575 9C23A020 4000168D     ....@..u.#. @...
.De
.IP
When used with the
.B \-v
option, the same line looks like:
.IP
.Ds
^A^@^@^@  @^@^U u 9C #A0    @^@^V8D     ....@..u.#. @...
.De
.\".TP
.\".B \-k
.\"`Kanji' option.
.\"Ordinarily,
.\".I most
.\"will go into binary mode if the file consists of non-ascii characters.
.\"Sometimes this feature is not desirable since some
.\"terminals have a special interpretation for eight bit
.\"characters.
.\"The
.\".B \-k
.\"option turns off the automatic sensing.
.TP
.B \-C
Disable color support.
.TP
.B \-d
Omit the backslash mark used to denote a wrapped line.
.TP
.B \-M
Disable the use of mmap.
.TP
.B \-s
Squeeze.
Replace multiple blank lines with a single blank line.
.TP
.B \-z
option turns off gunzip-on-the-fly.
.TP
.B \-v
Display control characters as in `^A' for control A.
Normally
.I most
does not interpret control characters.
.TP
.B \-t
Display tabs as `^I'.
This option is meaningful only when used with the
.B \-v
option.
.TP
.BI + lineno
Start up at
.IR lineno .
.TP
.B \-c
Make searches case sensitive.
By default, they are not.
.TP 
.B \-u
Disable UTF-8 mode even if the locale dictates it.
.TP
.B +u
Force UTF-8 mode.  By default most will use the current locale to
determine if UTF-8 mode shoul be used.  The
.B +u
and
.B \-u
switches allow the behavior to be overridden.
.TP
.B +d
This switch should only be used if you want the option to
delete a file while viewing it.
This makes it easier to
clean unwanted files out of a directory.
The file is
deleted with the interactive key sequence
.RB ` :D '
and then confirming with
.RB ` y '.
.TP
.BI +/ string
Start up at the line containing the first occurrence of
.IR string .
.SH "COMMAND USAGE"
The commands take effect immediately; it is not necessary to type a
carriage return.
.PP
In the following commands,
.I i
is a numerical argument
(1 by default).
.TP
.BR SPACE ", " CTRL-D ", " NEXT_SCREEN
Display another windowful, or jump
.I i
windowfuls if
.I i
is specified.
.TP
\fBRETURN\fR, \fBDOWN_ARROW\fR, \fBV\fR, \fBCTRL-N\fR
Display another line, or
.I i
more lines, if specified.
.TP
.BR UP_ARROW ", " ^ ", " CTRL-P
Display previous line, or
.I i
previous lines, if specified.
.TP
.BR T ", " ESCAPE <
Move to top of buffer.
.TP
.BR B ", " ESCAPE >
Move to bottom of buffer.
.TP
.BR RIGHT_ARROW ", " TAB ", " >
Scroll window left
.RI 60 i
columns to view lines that are beyond the right margin of the window.
.TP
.BR LEFT_ARROW ", " CTRL-B ", " <
Scroll window right
.RI 60 i
columns to view lines that are beyond the left margin of the window.
.TP
\fBU\fR, \fBCTRL-U\fR, \fBDELETE\fR, \fBPREV_SCREEN\fR
Skip back
.I i
windowfuls and then print a windowful.
.TP
.BR R ", " CTRL-R
Redraw the window.
.TP
.BR J ", " G
If
.I i
is not specified, then prompt for a line number then jump to that line
otherwise just jump to line
.IR i .
.TP
.B %
If
.I i
is not specified, then prompt for a
percent number then jump to that percent of the
file otherwise just jump to
.I i
percent of the file.
.TP
.BR W ", " w
If the current screen width is 80, make it 132 and vice-versa.
For other values, this command is ignored.
.TP
\fBQ\fR, \fBCTRL-X CTRL-C\fR, \fBCTRL-K E\fR
Exit from
.IR most .
On VMS, ^Z also exits.
.TP
.BR h ", " CTRL-H ", " HELP ", " PF2
Help.
Give a description of all the
.I most
commands.
The
.I most
environment variable
.B MOST_HELP
must be set for this to be meaningful.
.TP
\fBf\fR, \fB/\fR, \fBCTRL-F\fR, \fBFIND\fR, \fBGOLD PF3\fR
Prompt for a string and search forward from the
current line for
.IR i th
distinct line containing the string.
.B CTRL-G
aborts.
.TP
.B ?
Prompt for a string and search backward for the
.IR i th
distinct line containing the string.
.B CTRL-G
aborts.
.TP
.B n
Search for the next
.I i
lines containing an occurrence of the last search string in the
direction of the previous search.
.\"-------
.\" The '@' causes problems when included in a paragraph tag
.\" in my system's -man macro set, so jump though some hoops to
.\" avoid doing this.
.\"-------
.PP
\fBm\fR, \fBSELECT\fR, \fBCTRL-@\fR, \fBCTRL-K M\fR, \fBPERIOD\fR
.PD 0
.IP
.PD
Set a mark on the current line for later reference.
.TP
.BR "INSERT_HERE, CTRL-X CTRL-X, COMMA, CTRL-K RETURN, GOLD PERIOD"
Set a mark on the current line but return to previous mark.
This allows the user to toggle back and forth between two positions
in the file.
.TP
.BR l ", " L
Toggle locking for this window.
The window is locked if there is a `*' at the left edge
of the status line.
Windows locked together, scroll together.
.TP
.BR "CTRL-X 2" ", " "CTRL-W 2" ", " "GOLD X"
Split this window in half.
.TP
.BR "CTRL-X o" ", " "CTRL-W o" ", " o ", " GOLD UP ", " GOLD DOWN 
Move to other window.
.TP
.BR "CTRL-X 0" ", " "CTRL-W 0" ", " "GOLD V"
Delete this window.
.TP
.BR "CTRL-X 1" ", " "CTRL-W 1" ", " "GOLD O"
Delete all other windows, leaving only one window.
.TP
.BR E ", " e
Edit this file.
.TP
.BR $ ", " "ESC $"
This is system dependent.
On VMS, this causes
.I most
to spawn a subprocess.
When the user exits the process,
.I most
is resumed.
On
.UX
systems,
.I most
simply suspends itself.
.TP
.B :n
Skip to the next filename given in the command line.
Use the arrow keys to scroll forward or backward
through the file list.
.RB ` Q '
quits
.I most
and any other key selects the given file.
.TP
.B :c
Toggle case sensitive search.
.TP
.B :D
Delete current file.
This command is only meaningful with the
.B +d
switch.
.TP
.BR :o ", " :O
Toggle various options.
With this key sequence,
.I most
displays a prompt asking the user to hit
one of:
.BR bdtvw .
The
.RB ` b ',
.RB ` t ',
.RB ` v ',
and
.RB ` w '
options have the same meaning as the command
line switches.
For example, the
.RB ` w '
option will toggle wrapping on and off for the current window.
.IP
The
.RB ` d '
option must be used with a prefix integer
.IR i .
All lines indented beyond
.I i
columns will not be displayed.
For example, consider the fragment:
.IP
.Ds
.ne 11
int main(int argc, char **argv)
{
	int i;

	for (i = 0; i < argc, i++)
	{
		fprintf(stdout,"%i: %s\en",i,argv[i]);
	}
	return 0;
}
.De
.IP
The key sequence
.RB ` 1:od '
will cause
.I most
to display the file ignoring all lines indented beyond the first column.
So for the example above,
.I most
would display:
.IP
.Ds
.ne 2
int main(int argc, char **argv)...
}
.De
.IP
where the `...' indicates lines follow are not displayed.
.SH HINTS
.B CTRL-G
aborts the commands requiring the user to type something
in at a prompt.
The backquote key has a special meaning here.
It is used to quote certain characters.
This is useful when search for
the occurrence of a string with a control character or a string at
the beginning of a line.
In the latter case, to find the occurrence
of `The' at the beginning of a line, enter
.B `^JThe
where
.B `
quotes the
.BR CTRL-J .
.SH ENVIRONMENT
.I most
uses the following environment variables:
.TP
.B MOST_SWITCHES
This variable sets commonly used switches.
For example,
some people prefer to use
.I most
with the
.B \-s
option so that excess blank lines are not displayed.
On VMS this is normally done done in the login.com through the line:
.IP
.Ds
$ define MOST_SWITCHES "-s"
.De
.TP
.BR MOST_EDITOR ", " SLANG_EDITOR
Either of these environment variables specify an editor for
.I most
to invoke to edit a file. The value can contain %s and %d formatting
descriptors that represent the file name and line number,
respectively.  For example, if JED is your editor, then set
.B MOST_EDITOR 
to 'jed %s -g %d'.
.TP
.B MOST_HELP
This variable may be used to specify an alternate help file.
.TP
.B MOST_INITFILE
Set this variable to specify the initialization file to load during
startup.  The default action is to load the system configuration file
and then a personal configuration file called
.I .mostrc
on Unix, and
.I most.rc
on other systems.
.SH CONFIGURATION FILE SYNTAX
When most starts up, it tries to read a system configuration file and
then a personal configuration file.  These files may be used to
specify keybindings and colors.
.PP
To bind a key to a particular function use the syntax:
.PP
.B    setkey  function-name  key-sequence
.PP
The 
.I setkey
command requires two arguments.  The
.I function-name
argument specifies the function that is to be executed as a response
to the keys specified by the
.I key-sequence 
argument are pressed.  For example,
.PP
      setkey   "up"     "^P"
.PP
indicates that when
.I Ctrl-P
is pressed then the function
.I up
is to be executed.  
.PP
Sometimes, it is necessary to first unbind a key-sequence before
rebinding it in order via the
.I unsetkey
function:
.PP
       unsetkey "^F"
.PP
Colors may be defined through the use of the 
.I color

changes.txt
Cangees since 5.0.0a
1. src/most.h: Fixed slang v2.x supporting its own export of SLsnprintf.

Changes since 5.0.0
1. src/Makefile.in: Install docs in $prefix/share/doc

Changes since 4.10.2
0. src/most.c: version bumped to 5.0.0-x
1. src/line.c: Added support for embedded ANSI escape sequences that
   seem to be more common in man pages.
2. When switching searching directions, and no search string was
   entered at the search prompt, use the previous string.
3. Added patches for UTF-8 support kindly provided by Javier Kohen.
4. src/most.c: Removed references to Kanji from the usage message.
   Kanji is not supported by slang-2.  Use Unicode.
5. src/file.c: Most will automatically start up in binary mode if the
   file contains a nul-byte in the first 512 characters.  Previously,
   it would switch into binary mode when the file contained any
   characters with the most significant bit set.
6. autoconf/config.sub,config.guess: Updated to more recent versions.
7. src/most.c: Add +u/-u switches to force UTF-8 on/off.  The default
   behavior depends upon the locale.
8. src/sysdep.c: In most_getkey, if SLang_getkey returns an error,
   then just exit.
9. src/search.c: Added `#undef SLANG_REGEXP' and `#undef
   HAVE_V8_REGCOMP' until the regexp searches are rewritten.
10. src/edit.c: Filenames are quoted before being passed to the shell
    for editing.
11. src/file.c: Filenames are quoted before being passed to the unzip
    program.
12. src/sysdep.c: Check the process group id of the terminal and
    if it does not match that of most, then exit.
13. src/keym.c: Added "find_next_other_dir" function and bound it to N.
14. src/most.c: New commandline switch: "-d".  This switch disables
    the display of the \ character used to denote a wrapped line.
15. src/line.c: Added support for embedded ansi-color escape
    sequences, e.g., `ls --color | most`
16. src/window.c: Added an error hook to prevent slang from writing to
    stderr.
17. src/line.c: An ascii character+combining character was not being
    handled.

Changes since 4.10.1
1. src/window.c:most_read_from_minibuffer: An variable was not being
   initialized. (Kevin Oberman, oberman at es, net).
2. configure: mmap is enabled by default.  Use the --disable-mmap
   configure option of you do not want it.

Changes since 4.10.0
1. src/most.h: Prototype for SLang_set_error inconsistent with the
   function in most.c.
2. src/most.h: If compiled against slang2 without an snprintf
   function, use _pSLsnprintf.

Changes since 4.9.5
0. src/*.c: Various changes to support slang 2.  Version bumped to
   4.10.0.  UTF-8 support not yet available.
1. src/keym.c: Ignore mouse events under DOS/Windows (Gisle Vanem,
   giva at bgnett, no).

Changes since 4.9.4
1. src/line.c: A "length" variable was declared as unsigned char
   instead of unsigned int, causing a SEGV for columns larger than 255.
2. src/file.c: fixed a buffer overflow caused by a long file name on
   the command line.

Changes since 4.9.3

1. src/buffer.c: SLtt_Screen_Rows and SLtt_Screen_Cols were swapped
   causing a file to be improperly displayed in wrap mode.  Adrian P.
   van Bloois (adrian at accu.uu.nlhas) has my thanks for find this bug.
2. src/most.c: when run without arguments, S-Lang library version also
   printed.

Changes since 4.9.2

1. src/keym.c: added support for \xE0 type arrow keys (Win9X).
2. src/file.c: When resyncing mmaped buffer, make sure to preserve the
   current line number.
3. src/sysdep.c: SLang_getkey_intr_hook is unix
   specific and should not be used on VMS.
4. src/file.c: Make arrow-keys for next_file_cmd work on win32 systems.
5. src/file.h: be sure to quote filename argument (Mako Hill
   <mako@debian.org>)
6. src/file.c: added support for bzip2 files
7. src/*.c: replace calls to sprintf with _SLsnprintf (Steve Kemp
   <skx@tardis.ed.ac.uk>).

Changes since 4.9.1
0. Added support for using mmap to read files.  Use configure
   --enable-mmap to enable this experimental feature.
1. Wrapping bug fixed.  Also some optimizations were made to speed up
   the display of wrapped files.
2. File default.rc renamed to most.rc.
3. A^H_ interpreted as underline.  Previously, _^HA was recognized.
4. Jerome Lauret <JLAURET@mail.chem.sunysb.edu> provided most.hlp, as
   well as a few VMS patches.
5. src/Makefile.in: patch to make most's Makefile respect DESTDIR as
   well as removing the hardcoded use of $(prefix)/etc for the
   location of most's config file. <brad@comstyle.com>
6. src/line.c: some simplifications and cleanups designed to avoid
   nasty problems with extremely long lines.

Changes since 4.9.0
1. MOST_SWITCHES env variable can now contain extended switches.
2. If most is reading from stdin, it will reinit the terminal when it
   encounters an EOF on stdin.  This is to work around a problem under
   linux where for some reason, some other process resets the terminal.
Changes since 4.8.1
1. Apparantly, the binary flag +b is not valid with POSIX popen.
2. Screen resize code modified so that most works in a SunOS cmdtool.
3. Color support added.  Use `most -C' to disable.
4. Ported to Mingw32 (GNU C)

Changes since 4.7
1. Added tail -f type mode.  Use F to activate.
2. OS/2 port provided by j.burnell@irl.cri.nz (John Burnell).
3. Left/Right arrow scroll 1 column.  Use <, > to scroll in larger
   chunks.

Changes since 4.6
0. Various cleanups and other improvements
1. Binary mode displays byte offset
2. Ported to DJGPP

Changes since 4.5
1. Many changes to enable MOST use the S-Lang screen management routines.

Features in 4.5
1. More cleanup/bug fixes
2. Regular expression searches 
3. Automatic detection of compressed files.

Features in 4.41:
1. Modified to read raw directry file in BINARY mode.

New features in MOST 4.4:

1. User definable keymaps
2. ISO Latin1 8 bit clean
3. Edit and recall previous search strings
4. Much friendlier, lots of messages and help
5. VMS ``Gopher ready''
6. Cleaned up alot and small bugs fixed

src/most.c
/*
 This file is part of MOST.

 most 5.0.0b and beyond are also:
 Copyright (c) 2010 Kaity G. B. <uberChick@uberChicGeekChick.com>
 Released under the terms of the RPL 1.5 or above.

 most up to and including versions 5.0.0a are:
 Copyright (c) 1991, 1999, 2002, 2005, 2006, 2007 John E. Davis <ftp://space.mit.edu/pub/davis/most/>
 Relesed under the terms of the GPL 2 or above.

Unless explicitly acquired and licensed from Licensor under another
license, the contents of this file are subject to the Reciprocal Public
License ("RPL") Version 1.5, or subsequent versions as allowed by the RPL,
and You may not copy or use this file in either source code or executable
form, except in compliance with the terms and conditions of the RPL.

All software distributed under the RPL is provided strictly on an "AS
IS" basis, WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED, AND
LICENSOR HEREBY DISCLAIMS ALL SUCH WARRANTIES, INCLUDING WITHOUT
LIMITATION, ANY WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, QUIET ENJOYMENT, OR NON-INFRINGEMENT. See the RPL for specific
language governing rights and limitations under the RPL.
*/


#include "config.h"

#include <stdio.h>
#include <ctype.h>

#ifdef HAVE_STDLIB_H
# include <stdlib.h>
#endif

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#ifdef VMS
# include <rmsdef.h>
#endif

#include <errno.h>
#include <stdarg.h>

#include <slang.h>
#include "version.h"

#include "slang-macros.h"

#include "most.h"
#include "search.h"
#include "window.h"
#include "file.h"
#include "sysdep.h"
#include "keym.h"
#include "display.h"
#include "line.h"

int Most_S_Opt = 0;		       /* squeeze liness */
int Most_A_Opt = 1;		       /* automatically choose -b if necessary */
int Most_V_Opt = 0;		       /* display control chars */
int Most_B_Opt = 0;		       /* display Binary File */
int Most_T_Opt = 0;		       /* display tab as ^I-- valid only with V option */
int Most_D_Opt = 0;		       /* delete file mode  (see ':D')  */
int Most_K_Opt = 0;		       /* Display 8 bit unformatted (Kanji) */
int Most_Z_Opt = 0;		       /* Gunzip on the fly */
int Most_Want_Exit;
int Most_Secure_Mode;
int Most_Captive_Mode;
#if MOST_HAS_MMAP
int Most_Disable_MMap	= 0;
#endif

int Most_UTF8_Mode = -1;	       /* -1:auto, 0:off, 1:on */

static int  Most_Starting_Line;
char *Most_Program;	/* Program Name (argv[0]) */

static char *Most_Version = MOST_VERSION_STR;

#ifdef VMS
# ifndef isalpha
#  define isalpha(x) \
   (((x >= 'A') && (x <= 'Z'))||((x >= 'a') && (x <= 'z')) ? 1 : 0)
# endif
#endif

void most_usage (void)
{
   fprintf(stderr,"MOST version %s (S-Lang version %s)\n",
	   Most_Version, SLang_Version_String);
   if (SLang_Version != SLANG_VERSION)
     fprintf (stderr, " *Note: This executable was compiled against S-Lang %s\n", SLANG_VERSION_STRING);
     
   fprintf (stderr, "Usage:\n");
   fprintf(stderr, "most [-1Cbcdkstvw] [+/string] [+line number] [+s] [+d] file...\n");
   fputs(" where: -1:  assume VT100 terminal. (VMS only)\n", stderr);
   fputs("        -b:  Startup in binary mode.\n", stderr);
   fputs("        -C:  disable color support\n", stderr);
   fputs("        -c:  Make searches case sensitive.\n", stderr);
   fputs("        -d:  Do not display the \\ wrap marker when wrapping lines.\n", stderr);
   /* fputs("        -k:  Kanji mode.\n", stderr); */
#if MOST_HAS_MMAP
   fputs("        -M:  Do not attempt to mmap files.\n", stderr);
#endif
   fputs("        -s:  Squeeze out excess blank lines.\n", stderr);
   fputs("        -t:  Display tabs as ^I.  If this option is immediately followed\n", stderr);
   fputs("               by an integer, the integer sets the tab width.\n", stderr);
   fputs("        -u:  Disable UTF-8 mode\n", stderr);
   fputs("        -v:  Do not interpret backspace formatting characters.\n", stderr);
   fputs("        -w:  Wrap lines.\n", stderr);
   fputs("        -z:  No gunzip-on-the-fly.\n", stderr);
   fputs("        +/string:\n", stderr);
   fputs("             Search for string\n", stderr);
   fputs("        +line number\n", stderr);
   fputs("             Start up at specified line number.\n", stderr);
   fputs("        +d:  Allow file deletion.\n", stderr);
   fputs("        +s:  Secure Mode-- no edit, cd, shell, and reading files not\n", stderr);
   fputs("               already listed on the command line.\n", stderr);
   fputs("        +u:  Enable UTF-8 mode.\n", stderr);
   fprintf(stderr, "\nExample: most -ct4 +82 keymap.c\n");
   fputs(" makes searches case sensitive, sets tabwidth to 4, and displays the file\n", stderr);
   fputs(" keymap.c starting at line 82.\n", stderr);
}

static void do_switches(char *str);

static void do_extended_switches(char *str)
{
   int i;
   char ch;
   char numstr [256];

   i = 0;
   ch = *(++str);
   if ( ch == '/')
     {
	strcpy (Most_Search_Str,++str);
	return;
     }

   if (ch >= '0' && ch <= '9')
     {
	while ((ch >= '0' && ch <= '9') && (i < 10))
	  {
	     numstr[i++] = ch;
	     ch = *(++str);
	  }
	numstr[i] = '\0';
	if (1 == sscanf (numstr,"%d", &i))
	  Most_Starting_Line = i;
	return;
     }

   if (isalpha(ch))
     {
	while (1)
	  {
	     switch (ch)
	       {
		case 0:
		  return;
		case ' ':
		case '+':
		  break;
		case '-':
		  do_switches (str);
		  return;

		case 'D':
		case 'd':
		  Most_D_Opt = 1;   /* delete file mode */
		  break;
		case 'S':
		case 's':
		  Most_Secure_Mode = 1;
		  break;
		  
		case 'U':
		case 'u':
		  Most_UTF8_Mode = 1;  /* +u */
		  break;

		default:
		  fprintf(stderr,"%s invalid extended option %c ignored.\n",
			  Most_Program, ch);
	       }
	     ch = *(++str);
	  }
     }

   fprintf(stderr,"%s: switch '+%s' not valid.\n", Most_Program, str);
   exit (1);
}

/* if non-zero, assume terminal is only a generic vt100 */
static int assume_vt100 = 0;
static int No_Colors = 0;

static void do_switches(char *str)
{
   char ch;
   if (*str == '-') str++;
   while ((ch = *str++) != '\0')
     {
	switch (ch)
	  {
	   default:
	     fprintf(stderr,"%s: invalid switch %c ignored.\n",
		     Most_Program, ch);
	     break;

	   case 'C':
	     No_Colors = 1;
	     break;
	   case 'c':
	     Most_Case_Sensitive = 1;
	     break;
	   case 'd':
	   case 'D':
	     Most_Show_Wrap_Marker = 0;
	   case 's':
	   case 'S':
	     Most_S_Opt = 1; break;
	   case 'V':
	   case 'v':
	     Most_V_Opt = 1;  /* verbose-- convert control chars to '^' 'ch' */
	     break;
	   case 'W':
	   case 'w':  Most_W_Opt = 1; break;

	   case 'K':		       /* Kanji option */
	   case 'k':
	     /* Most_K_Opt = 1; break; */
	     break;

	   case 'B':
	   case 'b':
	     Most_B_Opt = 1;  /* Binary display 8 bit */
	     break;

	   case 'M':
#if MOST_HAS_MMAP
	     Most_Disable_MMap = 1;
#endif
	     break;

	   case 'z':
	   case 'Z':
	     Most_Z_Opt = 1;  /* NO Gunzip-on-the-fly */
	     break;

	   case 't':
	   case 'T': /* expand tabs to '^I'; meaningful only with 'v' */
	     ch = *str;
	     if ((ch <= '9') && (ch >= '0'))
	       {
		  str++;
		  Most_Tab_Width = (int) ch - '0';
		  if (Most_Tab_Width == 0) Most_T_Opt = 1;
	       }
	     else Most_T_Opt = 1;
	     break;

	   case 'n': case 'N':
	      /* could be the Gopher Naive user switch --- ignored. */
	     break;
	   case '1': assume_vt100 = 1;
	     break;
	     
	   case 'u':
	   case 'U':
	     Most_UTF8_Mode = 0;       /* -u */
	     break;

	     /* Allow MOST_SWITCHES environment variable to contain + forms,
	      * e.g., "-sn+d" or "-s -n +d"
	      */
	   case ' ':
	   case '-':
	     break;
	   case '+':
	     do_extended_switches (str - 1);   /* include '+' */
	     return;
	  }
     }
}

void most_exit_error(char *fmt,...)
{
   va_list ap;

   most_reset_tty ();
   most_reset_display();
   if (fmt != NULL)
     {
	va_start (ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end (ap);
	putc ('\n', stderr);
     }
#ifdef MALLOC_DEBUG
   SLmalloc_dump_statistics ();
#endif
   exit(1);
}

static void play_cat(char *file)
{
   char buf[4096 * 4];
   int n;
   FILE *fp;

   if (file == NULL) fp = stdin;
   else
     {
	fp = fopen(file, "r");
	if (fp == NULL) return;
     }

   while ((n = fread(buf, 1, 4096 * 4, fp)) > 0)
     {
	int m;
	m = fwrite (buf, 1, n, stdout);
	if (m != n)
	  {
	     fprintf (stderr, "fwrite returned %d, errno = %d\n",
		      m, errno);
	     exit (1);
	  }
     }
}

void most_initialize_most (void)
{
   Most_S_Opt = 0;
   Most_A_Opt = 1;
   Most_V_Opt = 0;
   Most_B_Opt = 0;
   Most_T_Opt = 0;
   Most_D_Opt = 0;
   Most_K_Opt = 0;
   Most_W_Opt = 0;

   Most_Selective_Display = 0;
   *Most_Search_Str = 0;   Most_Search_Dir = 1;
   Most_Top_Win = Most_Win = NULL;
   Most_Buf = NULL;
   Most_Eob = NULL;
   Most_Beg = NULL;
   Most_Captive_Mode = Most_Secure_Mode = 0;
   Most_Want_Exit = 0;
}

static void do_most (char *file, int start)
{
   int piped;
   MOST_INT row, col;

   most_get_cdir(Most_C_Dir);

   row = col = 0;
   if (file[0] == '\0') piped = 1; else piped = 0;

   if ((-1 == most_find_file (file))
       && (Most_Num_Files == 1))
     most_exit_error ("%s: failed to open for reading.", file);

   most_init_display ();

   most_goto_line(start);

   Most_Curs_Offset = Most_C_Offset;

   if (*Most_Search_Str
       && ((row = most_search (Most_Beg + Most_C_Offset, 1, &col)) > 0))
     most_goto_line(row);
   else
     {
	row = Most_C_Line;
	col = 1;
     }

   most_window_buffer();
   Most_Curs_Row = Most_Win->curs_line = row - Most_C_Line + 1;
   Most_Win->curs_offset = Most_Curs_Offset;
   Most_Curs_Col = Most_Win->curs_col = col;
   most_redraw_window();
   most_update_status();

   while (Most_Want_Exit == 0)
     {
	most_execute_key();
     }
}

void most_exit_most (void)
{
   if (Most_Want_Exit) return;
   Most_Want_Exit = 1;
   most_clear_minibuffer ();
   most_reset_tty ();
   most_reset_display ();
   most_free_windows ();
#ifdef MALLOC_DEBUG
   SLmalloc_dump_statistics ();
#endif
}

static void utf8_config (void)
{
   int utf8_mode = Most_UTF8_Mode;

   utf8_mode = SLutf8_enable (-1);     /* returns 0 or 1 */
   if (Most_UTF8_Mode == -1)
     Most_UTF8_Mode = utf8_mode;
   else if (utf8_mode != Most_UTF8_Mode)
     {
	if (utf8_mode == 1)
	  (void) SLsmg_utf8_enable (0);   /* locale is UTF-8, but -u passed */
	else
	  (void) SLsmg_utf8_enable (1);   /* locale not UTF-8, but +u passed */
     }
}

int most (int argc, char **argv)
{
   char file[MAX_PATHLEN], *switches;
   int file_i = 0, quit,i,piped,a_opt;
   unsigned long context;
   int status = 0;

#ifdef VMS
   char filename[256];
#else
   int j;
#endif

   Most_Program = argv[0];
   piped = 0;

   switches = getenv ("MOST_PROMPT");
   if ((switches != NULL) && (*switches != 0)) Most_Global_Msg = switches;

   switches = getenv("MOST_SWITCHES");
   if (switches !=  NULL)  do_switches(switches);

   i = 1;
   if (argc > 1)
     {
	quit = 0;
	while ((!quit) && (i < argc))
	  {
	     if (argv[i][0] == '-')
	       do_switches(argv[i++]);
	     else if (argv[i][0] == '+')
	       do_extended_switches(argv[i++]);
	     else quit = 1;
	  }
     }

#if MOST_HAS_MMAP
   /* if (Most_D_Opt) */
   /*   Most_Disable_MMap = 1; */
#endif

   if (i == argc)
     {
	if (isatty(0))   /* 1 if stdin is a terminal, 0 otherwise */
	  {
	     most_usage ();
	     return 0;
	  }
	/* assume input is from stdin */
	file[0] = '\0';  /* tells most this is stdin */
	piped = 1;
	if (!isatty(fileno(stdout)))
	  {
	     play_cat(NULL);
	     return 0;
	  }
     }
   else
     {
	strncpy (file, argv[i], sizeof(file));
	file[sizeof(file)-1] = 0;
     }

   if (!isatty(fileno(stdout)))
     {
	while (i < argc) play_cat(argv[i++]);
	exit(0);
     }

   Most_Num_Files = 0;
   context = 0;

   SLtt_get_terminfo();
   utf8_config ();
   SLtt_Ignore_Beep = 1;
   if (No_Colors) 
     SLtt_Use_Ansi_Colors = 0;

   most_setup_colors ();
   most_init_tty ();
   most_init_keymaps ();

   if (Most_B_Opt) Most_A_Opt = 0;   /* explicit b overrides a */
   a_opt = Most_A_Opt;

   if (!piped)
     {
	file_i = i;
#ifdef VMS
	while(i < argc)
	  {
	     if (Most_Num_Files >= MOST_MAX_FILES) break;
	     if (argv[i][0] == '.') strcpy(file,"*"); else *file = 0;
	     strcat(file, most_unix2vms(argv[i++]));
	     while (RMS$_NORMAL == (status = most_expand_file_name(file,filename)))
	       {
		  Most_File_Ring[Most_Num_Files] = (char*) MOSTMALLOC(strlen(filename) + 1);
		  strcpy(Most_File_Ring[Most_Num_Files++], filename);
	       }
	     if (status == RMS$_NMF) status = RMS$_NORMAL; /* avoid spurious warning message */
	  }
	
	if (Most_Num_Files) strcpy(file,Most_File_Ring[0]);
	else fputs("%%MOST-W-NOFILES, no files found\n", stderr);
#else
	Most_Num_Files = argc - i;
	if (Most_Num_Files > MOST_MAX_FILES)
	  {
	     Most_Num_Files = MOST_MAX_FILES;
	     argc = Most_Num_Files + i;
	  }

	j = 0;
	while (i < argc)
	  {
	     Most_File_Ring[j++] = argv[i++];
	  }
#endif
     }

   if (Most_Num_Files || piped) do_most(file, Most_Starting_Line);
   else if (Most_Num_Files == 0)
     fprintf(stderr,"File %s not found\n", argv[file_i]);

   most_exit_most ();
   return status;
}

#if SLANG_VERSION <= 10409

int SLang_set_error (int x)
{
   SLang_Error = x;
   return 0;
}

int SLang_get_error (void)
{
   return SLang_Error;
}

#endif
